### PR TITLE
[static_routes] bug fixes and optimisation 

### LINKED ISSUE #782: ios_static_routes is not idempotent with IPv6 static routes that have a lowercase next-hop
##### SUMMARY

When the cisco.ios.ios_static_routes module configures an IPv6 static route that has a next-hop IPv6 address (`forward_router_address`) that contains lowercase alphabetical characters, the IPv6 static route is configured as expected, but subsequent runs of the module are not idempotent.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

ios_static_routes

##### ANSIBLE VERSION

```paste below
ansible [core 2.14.3]
  config file = /home/chart2/foo/ansible/ansible.cfg
  configured module search path = ['/home/chart2/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/chart2/foo/venv/lib/python3.10/site-packages/ansible
  ansible collection location = /home/chart2/foo/ansible/collections
  executable location = /home/chart2/foo/venv/bin/ansible
  python version = 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] (/home/chart2/foo/venv/bin/python3)
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION

```paste below
# /home/chart2/foo/ansible/collections/ansible_collections
Collection Version
---------- -------
cisco.ios  4.4.0  
```

##### CONFIGURATION

```paste below
COLLECTIONS_PATHS(/home/chart2/foo/ansible/ansible.cfg) = ['/home/chart2/foo/ansible/collections']
CONFIG_FILE() = /home/chart2/foo/ansible/ansible.cfg
DEFAULT_FORKS(/home/chart2/foo/ansible/ansible.cfg) = 100
DEFAULT_HOST_LIST(/home/chart2/foo/ansible/ansible.cfg) = ['/home/chart2/foo/ansible/inventory.yaml']
DEFAULT_ROLES_PATH(/home/chart2/foo/ansible/ansible.cfg) = ['/home/chart2/.ansible/roles', '/usr/share/ansible/roles', '/etc/ansible/roles']
HOST_KEY_CHECKING(/home/chart2/foo/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT

Observed when executed against a Catalyst 8000v running IOS-XE 17.06.01a.

##### STEPS TO REPRODUCE

Execute the playbook below against a device.

```yaml
---
- name: "Configure IPv6 addresses on devices"
  hosts: endpoints
  gather_facts: false
  tasks:
    - name: "Configure IPv6 default routes on endpoints"
      cisco.ios.ios_static_routes:
        config:
          - address_families:
              - afi: ipv6
                routes:
                  - dest: ::/0
                    next_hops:
                      - forward_router_address: "2001:db8:1::1"
      become: true
```

On the first execution, the module changes the device's configuration as expected.

```
PLAY [Configure IPv6 addresses on devices] *******************************************************************************************************************************************************************

TASK [Configure IPv6 default routes on endpoints] ************************************************************************************************************************************************************
changed: [Virtual-Host-1]

PLAY RECAP ***************************************************************************************************************************************************************************************************
Virtual-Host-1             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

The device configuration is shown below:

```
Virtual-Host-1#show running-config | include ipv6 route
ipv6 route ::/0 2001:DB8:1::1
```

Then, re-run the playbook. The output of the module indicates the device's configuration was changed again, even though the IPv6 static route is already present on the device.

```
PLAY [Configure IPv6 addresses on devices] *******************************************************************************************************************************************************************

TASK [Configure IPv6 default routes on endpoints] ************************************************************************************************************************************************************
changed: [Virtual-Host-1]

PLAY RECAP ***************************************************************************************************************************************************************************************************
Virtual-Host-1             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

This is resolved if alphabetical characters in the next-hop IPv6 address are capitalized.

```yaml
---
- name: "Configure IPv6 addresses on devices"
  hosts: endpoints
  gather_facts: false
  tasks:
    - name: "Configure IPv6 default routes on endpoints"
      cisco.ios.ios_static_routes:
        config:
          - address_families:
              - afi: ipv6
                routes:
                  - dest: ::/0
                    next_hops:
                      - forward_router_address: "2001:DB8:1::1"
      become: true
```

Run the modified playbook, and the module reports no changes needed.

```
PLAY [Configure IPv6 addresses on devices] *******************************************************************************************************************************************************************

TASK [Configure IPv6 default routes on endpoints] ************************************************************************************************************************************************************
ok: [Virtual-Host-1]

PLAY RECAP ***************************************************************************************************************************************************************************************************
Virtual-Host-1             : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

##### EXPECTED RESULTS

I expect the module to be idempotent, regardless of whether next-hop IPv6 addresses include uppercase or lowercase (or both!) alphabetical characters.

##### ACTUAL RESULTS

```paste below
ansible-playbook [core 2.14.3]
  config file = /home/chart2/foo/ansible/ansible.cfg
  configured module search path = ['/home/chart2/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/chart2/foo/venv/lib/python3.10/site-packages/ansible
  ansible collection location = /home/chart2/foo/ansible/collections
  executable location = /home/chart2/foo/venv/bin/ansible-playbook
  python version = 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] (/home/chart2/foo/venv/bin/python3)
  jinja version = 3.1.2
  libyaml = True
Using /home/chart2/foo/ansible/ansible.cfg as config file
setting up inventory plugins
host_list declined parsing /home/chart2/foo/ansible/inventory.yaml as it did not pass its verify_file() method
script declined parsing /home/chart2/foo/ansible/inventory.yaml as it did not pass its verify_file() method
Parsed /home/chart2/foo/ansible/inventory.yaml inventory source with yaml plugin
Loading collection cisco.ios from /home/chart2/foo/ansible/collections/ansible_collections/cisco/ios
redirecting (type: action) cisco.ios.ios_static_routes to cisco.ios.ios
Loading callback plugin default of type stdout, v2.0 from /home/chart2/foo/venv/lib/python3.10/site-packages/ansible/plugins/callback/default.py
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: test2.yaml *****************************************************************************************************************************************************************************************
Positional arguments: test2.yaml
verbosity: 4
connection: smart
timeout: 10
become_method: sudo
tags: ('all',)
inventory: ('/home/chart2/foo/ansible/inventory.yaml',)
subset: Virtual-Host-1
forks: 100
1 plays in test2.yaml

PLAY [Configure IPv6 addresses on devices] *******************************************************************************************************************************************************************
redirecting (type: action) cisco.ios.ios_static_routes to cisco.ios.ios
Loading collection ansible.netcommon from /home/chart2/foo/ansible/collections/ansible_collections/ansible/netcommon

TASK [Configure IPv6 default routes on endpoints] ************************************************************************************************************************************************************
task path: /home/chart2/foo/ansible/test2.yaml:6
Loading collection ansible.utils from /home/chart2/foo/ansible/collections/ansible_collections/ansible/utils
<10.12.118.241> attempting to start connection
<10.12.118.241> using connection plugin ansible.netcommon.network_cli
Found ansible-connection at path /home/chart2/foo/venv/bin/ansible-connection
<10.12.118.241> local domain socket does not exist, starting it
<10.12.118.241> control socket path is /home/chart2/.ansible/pc/1003507cb9
<10.12.118.241> Loading collection ansible.netcommon from /home/chart2/foo/ansible/collections/ansible_collections/ansible/netcommon
<10.12.118.241> Loading collection ansible.utils from /home/chart2/foo/ansible/collections/ansible_collections/ansible/utils
<10.12.118.241> Loading collection cisco.ios from /home/chart2/foo/ansible/collections/ansible_collections/cisco/ios
<10.12.118.241> local domain socket listeners started successfully
<10.12.118.241> loaded cliconf plugin ansible_collections.cisco.ios.plugins.cliconf.ios from path /home/chart2/foo/ansible/collections/ansible_collections/cisco/ios/plugins/cliconf/ios.py for network_os cisco.ios.ios
<10.12.118.241> ssh type is set to auto
<10.12.118.241> autodetecting ssh_type
<10.12.118.241> ssh type is now set to libssh
<10.12.118.241> 
<10.12.118.241> local domain socket path is /home/chart2/.ansible/pc/1003507cb9
redirecting (type: action) cisco.ios.ios_static_routes to cisco.ios.ios
redirecting (type: action) cisco.ios.ios_static_routes to cisco.ios.ios
<10.12.118.241> ANSIBLE_NETWORK_IMPORT_MODULES: enabled
<10.12.118.241> ANSIBLE_NETWORK_IMPORT_MODULES: found cisco.ios.ios_static_routes  at /home/chart2/foo/ansible/collections/ansible_collections/cisco/ios/plugins/modules/ios_static_routes.py
<10.12.118.241> ANSIBLE_NETWORK_IMPORT_MODULES: running cisco.ios.ios_static_routes
<10.12.118.241> ANSIBLE_NETWORK_IMPORT_MODULES: complete
changed: [Virtual-Host-1] => {
    "after": [
        {
            "address_families": [
                {
                    "afi": "ipv4",
                    "routes": [
                        {
                            "dest": "0.0.0.0/0",
                            "next_hops": [
                                {
                                    "forward_router_address": "10.12.118.1"
                                }
                            ]
                        }
                    ]
                }
            ],
            "vrf": "management"
        },
        {
            "address_families": [
                {
                    "afi": "ipv6",
                    "routes": [
                        {
                            "dest": "::/0",
                            "next_hops": [
                                {
                                    "forward_router_address": "2001:DB8:1::1"
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ],
    "before": [
        {
            "address_families": [
                {
                    "afi": "ipv4",
                    "routes": [
                        {
                            "dest": "0.0.0.0/0",
                            "next_hops": [
                                {
                                    "forward_router_address": "10.12.118.1"
                                }
                            ]
                        }
                    ]
                }
            ],
            "vrf": "management"
        },
        {
            "address_families": [
                {
                    "afi": "ipv6",
                    "routes": [
                        {
                            "dest": "::/0",
                            "next_hops": [
                                {
                                    "forward_router_address": "2001:DB8:1::1"
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ],
    "changed": true,
    "commands": [
        "ipv6 route ::/0 2001:db8:1::1"
    ],
    "invocation": {
        "module_args": {
            "config": [
                {
                    "address_families": [
                        {
                            "afi": "ipv6",
                            "routes": [
                                {
                                    "dest": "::/0",
                                    "next_hops": [
                                        {
                                            "dhcp": null,
                                            "distance_metric": null,
                                            "forward_router_address": "2001:db8:1::1",
                                            "global": null,
                                            "interface": null,
                                            "multicast": null,
                                            "name": null,
                                            "permanent": null,
                                            "tag": null,
                                            "track": null
                                        }
                                    ],
                                    "topology": null
                                }
                            ]
                        }
                    ],
                    "vrf": null
                }
            ],
            "running_config": null,
            "state": "merged"
        }
    }
}

PLAY RECAP ***************************************************************************************************************************************************************************************************
Virtual-Host-1             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

### PR DESCRIPTION
**Pull request by KB-perByte**
##### SUMMARY

Fixes #782 #351 
The PR fixes a couple of under-mentioned issues
- [x] Fix idempotency with overridden and replaced state
- [x] Fix command generation ordering when next_hop contains interfaces
- [x] Fix config command generation for ipv4 and ipv6 static routes
- [x] Fix idempotency on upper case vs lower case ipv6 address

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

ios_static_routes

##### ADDITIONAL INFORMATION

Fix failure to process config like - 
```
ipv6 route vrf testVrf2 2001:DB8:0:3::/64 2001:DB8:0:3::2 tag 22 track 22 name name1
ip route 198.51.100.0 255.255.255.0 GigabitEthernet2 198.51.101.11 22 tag 22 permanent name nm1 multicast
```

```paste below
    # The full traceback is:
    #   File "/home/s..collections/ansible_collections/ansible/netcommon/plugins/module_utils/network/common/facts/facts.py", line 135, in get_network_resources_facts
    #     inst.populate_facts(
    #   File "/home/s..collections/ansible_collections/cisco/ios/plugins/module_utils/network/ios/facts/static_routes/static_routes.py", line 70, in populate_facts
    #     obj = self.render_config(self.generated_spec, key, same_dest[key])
    #   File "/home/s..collections/ansible_collections/cisco/ios/plugins/module_utils/network/ios/facts/static_routes/static_routes.py", line 200, in render_config
    #     if is_valid_ip(route[2]):
    # fatal: [35.155.113.92]: FAILED! => changed=false
    #   invocation:
    #     module_args:
    #       config: null
    #       running_config: null
    #       state: gathered
    #   msg: list index out of range

```

### CODE REVIEW

File: plugins/module_utils/network/ios/rm_templates/static_routes.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            "name": "ipv4",
+            "getval": re.compile(
+                r"""
+                ^ip\sroute
+                (\stopology\s(?P<topology>\S+))?
+                (\svrf\s(?P<vrf>\S+))?
+                (\s(?P<dest>\S+))
+                (\s(?P<netmask>\S+))
+                (\s(?P<interface>(ACR|ATM-ACR|Analysis-Module|AppNav-Compress|AppNav-UnCompress|Async|Auto-Template|BD-VIF|BDI|BVI|Bluetooth|CDMA-Ix|CEM-ACR|CEM-PG|CTunnel|Container|Dialer|EsconPhy|Ethernet-Internal|Fcpa|Filter|Filtergroup|GigabitEthernet|IMA-ACR|LongReachEthernet|Loopback|Lspvif|MFR|Multilink|NVI|Null|PROTECTION_GROUP|Port-channel|Portgroup|Pos-channel|SBC|SDH_ACR|SERIAL-ACR|SONET_ACR|SSLVPN-VIF|SYSCLOCK|Serial-PG|Service-Engine|TLS-VIF|Tunnel|VPN|Vif|Vir-cem-ACR|Virtual-PPP|Virtual-TokenRing)\d+))?
```
Comment (NilashishC) [inline plugins/module_utils/network/ios/rm_templates/static_routes.py:40]: Do we need to check for the interface type here?

File: plugins/module_utils/network/ios/rm_templates/static_routes.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        },
+        {
+            "name": "ipv6",
+            "getval": re.compile(
+                r"""
+                ^ipv6\sroute
+                (\stopology\s(?P<topology>\S+))?
+                (\svrf\s(?P<vrf>\S+))?
+                (\s(?P<dest>\S+))
+                (\s(?P<interface>(ACR|ATM-ACR|Analysis-Module|AppNav-Compress|AppNav-UnCompress|Async|Auto-Template|BD-VIF|BDI|BVI|Bluetooth|CDMA-Ix|CEM-ACR|CEM-PG|CTunnel|Container|Dialer|EsconPhy|Ethernet-Internal|Fcpa|Filter|Filtergroup|GigabitEthernet|IMA-ACR|LongReachEthernet|Loopback|Lspvif|MFR|Multilink|NVI|Null|PROTECTION_GROUP|Port-channel|Portgroup|Pos-channel|SBC|SDH_ACR|SERIAL-ACR|SONET_ACR|SSLVPN-VIF|SYSCLOCK|Serial-PG|Service-Engine|TLS-VIF|Tunnel|VPN|Vif|Vir-cem-ACR|Virtual-PPP|Virtual-TokenRing)\d+))?
```
Comment (NilashishC) [inline plugins/module_utils/network/ios/rm_templates/static_routes.py:96]: Same as above.

File: plugins/module_utils/network/ios/rm_templates/static_routes.py (hunk 1 above)
Comment (KB-perByte) [reply plugins/module_utils/network/ios/rm_templates/static_routes.py:40]: If we don't check the interface type it gets packed as dest or netmask,
`ip route 198.51.xx.x 255.255.255.0 GigabitEthernet2 198.51.yy.yy 22 tag 22 permanent name nm1 multicast `
^ this being a valid config line

File: plugins/module_utils/network/ios/rm_templates/static_routes.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            "name": "ipv4",
+            "getval": re.compile(
+                r"""
+                ^ip\sroute
+                (\stopology\s(?P<topology>\S+))?
+                (\svrf\s(?P<vrf>\S+))?
+                (\s(?P<dest>\S+))
+                (\s(?P<netmask>\S+))
+                (\s(?P<interface>(ACR|ATM-ACR|Analysis-Module|AppNav-Compress|AppNav-UnCompress|Async|Auto-Template|BD-VIF|BDI|BVI|Bluetooth|CDMA-Ix|CEM-ACR|CEM-PG|CTunnel|Container|Dialer|EsconPhy|Ethernet-Internal|Fcpa|Filter|Filtergroup|GigabitEthernet|IMA-ACR|LongReachEthernet|Loopback|Lspvif|MFR|Multilink|NVI|Null|PROTECTION_GROUP|Port-channel|Portgroup|Pos-channel|SBC|SDH_ACR|SERIAL-ACR|SONET_ACR|SSLVPN-VIF|SYSCLOCK|Serial-PG|Service-Engine|TLS-VIF|Tunnel|VPN|Vif|Vir-cem-ACR|Virtual-PPP|Virtual-TokenRing)\d+))?
+                (\s(?P<forward_router_address>(?!multicast|dhcp|global|tag|track|permanent|name)\S+))?
```
[the remark this answers is not in the document]
Comment (KB-perByte) [reply plugins/module_utils/network/ios/rm_templates/static_routes.py:41]: this again is a valid config
```
ipv6 route vrf testVrf2 2001:DB8:0:3::/64 GigabitEthernet2 tag 2 permanent name nm1
ipv6 route vrf testVrf2 2001:DB8:0:3::/64 Null0
ipv6 route vrf testVrf2 2001:DB8:0:3::/64 2001:DB8:0:3::3 22 multicast permanent name pp1
```

**Review by ashwini-mhatre — APPROVED** [review]

**Pull request merged**